proxy_ring_buffer.go

```diff
@@ -0,0 +1,278 @@
+ * limitations under the License.
+ */
+
+package hazelcast
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	hzcerrors "github.com/hazelcast/hazelcast-go-client/internal/hzerrors"
+	"github.com/hazelcast/hazelcast-go-client/internal/proto/codec"
+	iserialization "github.com/hazelcast/hazelcast-go-client/internal/serialization"
+)
+
+// A Ringbuffer is a data structure where the content is stored in a ring-like
+// structure. A Ringbuffer has a fixed capacity, so it won't grow beyond
+// that capacity and endanger the stability of the system. If that capacity
+// is exceeded, the oldest item in the Ringbuffer is overwritten.
+//
+// For details see https://docs.hazelcast.com/hazelcast/latest/data-structures/ringbuffer
+type Ringbuffer struct {
+	*proxy
+	partitionID int32
+	capacity    int64
+}
+
+// OverflowPolicy
+// Using this OverflowPolicy one can control the behavior what should to be done
+// when an item is about to be added to the RingBuffer, but there is {@code 0}
+// remaining capacity.
+//
+// Overflowing happens when a time-to-live is set and the oldest item in
+// the RingBuffer (the head) is not old enough to expire.
+type OverflowPolicy int
+type ReadResultSet struct {
+	rb        *Ringbuffer
+	readCount int32
+	items     []iserialization.Data
+	itemSeqs  []int64
+	nextSeq   int64
+}
+
+const (
+	// OverflowPolicyOverwrite
+	// Using this OverflowPolicyOverwrite policy the oldest item is overwritten
+	// no matter it is not old enough to retire. Using this policy you are
+	// sacrificing the time-to-live in favor of being able to write.
+	//
+	// Example: if there is a time-to-live of 30 seconds, the buffer is full
+	// and the oldest item in the ring has been placed a second ago, then there
+	// are 29 seconds remaining for that item. Using this policy you are going
+	// to overwrite no matter what.
+	OverflowPolicyOverwrite OverflowPolicy = 0
+
+	// OverflowPolicyFail
+	// Using this policy the call will fail immediately and the oldest item will
+	// not be overwritten before it is old enough to retire. So this policy
+	// sacrificing the ability to write in favor of time-to-live.
+	//
+	// The advantage of OverflowPolicyFail is that the caller can decide what to do
+	// since it doesn't trap the thread due to backoff.
+	//
+	// Example: if there is a time-to-live of 30 seconds, the buffer is full
+	// and the oldest item in the ring has been placed a second ago, then there
+	// are 29 seconds remaining for that item. Using this policy you are not
+	// going to overwrite that item for the next 29 seconds.
+	OverflowPolicyFail OverflowPolicy = 1
+
+	// ReadResultSetSequenceUnavailable is used when error happened
+	ReadResultSetSequenceUnavailable int64 = -1
+)
+
+func newRingbuffer(p *proxy) (*Ringbuffer, error) {
+	partitionID, err := p.stringToPartitionID(p.name)
+	if err != nil {
+		return nil, err
+	}
+	return &Ringbuffer{proxy: p, partitionID: partitionID, capacity: ReadResultSetSequenceUnavailable}, nil
```

Comment:
Could you layout this so each field is in its own line? Like:
```
rb := &Ringbuffer{
    proxy: p,
    partitionID: partitionID,
...
}
return rb, nil
```

Reply:
Done.
see 131f287231745e43d060a6bdc58078f9a1cc1dac

Reply:
And 05d7904de418415d8a016d7e12551b0a6c5d334c